product_tax_multicompany_default/models/product.py

```diff
@@ -0,0 +1,258 @@
+                propagate_taxes = True
+        else:
+            current_names = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).mapped("name")
+            current_tax_ids = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).ids
+            for company_id, value in taxes_bd.items():
+                if current_names == list(value.values()):
+                    continue
+                # We are looking to see if there are any taxes that can be applied to
+                # other companies from the taxes that the current product has.
+                candidates = self._taxes_by_company(
+                    field_name_data,
+                    self.env["res.company"].browse(company_id),
+                    current_tax_ids,
+                )
+                if len(candidates) == len(current_tax):
+                    propagate_taxes = True
+                    break
+
+        return propagate_taxes
+
+    def _get_product_taxes(self, field):
+        """Returns taxes on purchase or sales taxes
+
+        We need the product taxes for other companies, and we cannot take
+        them from the object itself, so we have to consult the database
+        where the product taxes passed by parameter are located for all
+        companies.
+
+        :param field: can be 'taxes' or 'purchase'
+        """
+        self.ensure_one()
+
+        if not self.id:
+            return []
+
+        table_by_field = {
+            "taxes": "product_taxes_rel",
+            "purchase": "product_supplier_taxes_rel",
+        }
+        table_name = table_by_field.get(field)
+
+        if not table_name:
+            raise ValueError("field should be 'taxes' or 'purchase'")
+
+        sql = f"""
+            SELECT DISTINCT a_tax.company_id, a_tax.name
+            FROM {table_name} ptr
+            LEFT JOIN account_tax a_tax ON tax_id = a_tax.id
+            WHERE ptr.prod_id = %s
+            """
+        self.env.cr.execute(sql, [self._origin.id])
+        return self.env.cr.fetchall()
+
+    def _taxes_by_company(self, field, company, match_tax_ids=None):
+        taxes_ids = []
+        if match_tax_ids is None:
+            taxes_ids = company[field].ids
+        # If None: return default taxes
+        if not match_tax_ids:
+            return taxes_ids
+        type_tax_use = "sale" if field == "account_sale_tax_id" else "purchase"
+        account_tax = self.env["account.tax"].sudo().browse(match_tax_ids)
+        taxes_ids = account_tax.filtered_domain([("type_tax_use", "=", type_tax_use)])
+        if not taxes_ids:
+            return taxes_ids
+        taxes_name = list({n for n in taxes_ids.mapped("name") if n})
+        taxes_ids = account_tax.search(
+            [
+                ("type_tax_use", "=", type_tax_use),
+                ("company_id", "=", company.id),
+                ("name", "in", taxes_name),
+            ]
+        ).ids
```

Comment:
type_tax_use has been already filtered
taxes_name will have the same data because comes from type_use_taxes

```suggestion
        return type_use_taxes.filtered_domain([("company_id", "=", company.id)]).ids
```

Reply:
We need to do a search for the taxes of other companies that are in the current record set.